crates/ruff_python_parser/src/parser/expression.rs

```diff
@@ -2777,18 +2784,28 @@ impl<'src> Parser<'src> {
             unreachable!()
         };
 
-        if !matches!(kind, IpyEscapeKind::Magic | IpyEscapeKind::Shell) {
-            // This should never occur as the lexer won't allow it.
-            unreachable!("IPython escape command expression is only allowed for % and !");
-        }
-
         let command = ast::ExprIpyEscapeCommand {
             range: self.node_range(start),
             node_index: AtomicNodeIndex::NONE,
             kind,
             value,
         };
 
+        // test_err ipython_help_escape_command_as_expression
+        // # parse_options: {"mode": "ipython"}
+        // with a,?b: pass
+        if !matches!(kind, IpyEscapeKind::Magic | IpyEscapeKind::Shell) {
+            // IPython escape commands like `?` (Help) are not allowed in expression contexts.
+            // Only `%` (Magic) and `!` (Shell) escape commands can be used as expressions.
+            // The lexer should prevent this, but we check here as a defensive measure.
+            self.add_error(
+                ParseErrorType::OtherError(format!(
+                    "IPython escape command expression is only allowed for % and !, found {kind:?}"
+                )),
+                &command,
+            );
+        }
```

Comment:
And this, you can move the test cases above the recovery branch in `parser/mod.rs`